### PR TITLE
Fix environmental issues related to python/pip/brew/npm etc.

### PR DESCRIPTION
**Pull request by xcodebuild**
- use homebrew to install pipenv on macOS, fix https://github.com/OpenDevin/OpenDevin/issues/544
- use `python3` if `python` is not available (e.g. on macOS)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -9,21 +9,35 @@ DEFAULT_WORKSPACE_DIR = "./workspace"
 DEFAULT_MODEL = "gpt-4-0125-preview"
 CONFIG_FILE = config.toml
 
+UNAME_SYS = $(shell uname -s)
+# fallback to python3 if python is not available (e.g. on macOS)
+PYTHON_BIN = $(shell which python3 || which python)
```
Comment (foragerr) [inline Makefile:14]: The comment says use python as default, and if that isn't available use python3
The code line does the opposite, it defaults to python3.

Also what should happen if neither is available?

Comment (xcodebuild) [reply Makefile:14]: My question, I plan to use `python` by default. Perhaps I should consider adding a check for when both `python3` and `python` are not available.

**Comment by spuder** [comment]
Won't this require that the user already have homebrew installed? 
I'm not sure it ubiquitous enough to be an assumed dependency.

**Comment by rbren** [comment]
I feel like we should try for `pip install pipenv` before falling back to homebrew. But maybe it doesn't matter?

**Comment by xcodebuild** [comment]
> As someone who constantly has a broken python install on OSX I appreciate this PR.
> 
> Won't this require that the user already have homebrew installed? I'm not sure it ubiquitous enough to be an assumed dependency.
> 
> Indirectly Related: #481

Good suggestion, currently trying to check for the existence of `brew` before use, if not, remind the user to install.

**Comment by xcodebuild** [comment]
I think it makes sense to do this, the current overall logic is:
1. Check whether python or python3 exists, prioritize python and fallback to python3 if neither exists, then give an error
1. Check if pipenv exists, if it does, skip it
2. macOS users try to install pipenv with `pip`, if it fails, fallback to use `homebrew` (if the current user has not installed `homebrew`, it will fail and remind the user)
3. Other operating systems still use `pip` for installation

**Comment by xcodebuild** [comment]
Added additional detection for `npm` and `nodejs` for https://github.com/OpenDevin/OpenDevin/issues/634

File: Makefile [hunk 2]
```diff
@@ -9,21 +9,54 @@ DEFAULT_WORKSPACE_DIR = "./workspace"
...
 # Build
 build:
 	@echo "Building project..."
 	@echo "Pulling Docker image..."
 	@docker pull $(DOCKER_IMAGE)
 	@echo "Installing Python dependencies..."
-	@python -m pip install pipenv
-	@python -m pipenv install -v
+	@$(INSTALL_PIPENV_CMD)
+	@$(PYTHON_BIN) -m pipenv install
```
Comment (subos2008) [inline Makefile:52]: With a brew installed pipenv on MacosX this gives me:

> $  python3 -m pipenv install
/usr/local/opt/python@3.12/bin/python3.12: No module named pipenv

`pipenv` is in the path and can be called directly.

Comment (Jordan-88) [reply Makefile:52]: I could not use the brew version of python 3.12 myself I had to install from python.org

**Comment by touhidurrr** [comment]
My wsl is saying cannot find python. Had to find and replace `python` to `python3` to fix this.
Maybe do something like `alias python=python3` to fix this? (Did not work when I added this line in makefile for some reason)

**Comment by dorbanianas** [comment]
can you pls pull the new main and try?

**Comment by xcodebuild** [comment]
> can you pls pull the new main and try?

The main branch has already switched to poetry, it seems this PR is no longer necessary, I'll close it for now.

**Comment by dorbanianas** [comment]
Enjoy! if you faced any issues let me know 👌

**Comment by touhidurrr** [comment]
> Enjoy! if you faced any issues let me know 👌

Now there is no python not found error. But the build is never finishing. My laptop is running out of storage because this project is installing gigabytes of random nvidia packages for no reason as I only intend to try out opendevin with my OpenAI keys and my laptop has no external gpu to begin with (using APU).

**Comment by dorbanianas** [comment]
Can you create a new issue for that pls I can try to figure it out

**Pull request closed without merging**